Support metadata for pyinfra plugins.

rod7760 opened the pull request:
## Changes
* Add pyinfra-metadata-schema-1.0.0.json.
* Add pyinfra-metadata.toml with current plugins.
* Add Jinja2 templating to sphinx build.
* Extended build-public-docs.sh to build local docs.
* Added generated index for facts and operations.

## Bench Test
```
scripts/build-public-docs.sh 
```

## Screenshot of Change
<img width="2872" height="1468" alt="image" src="https://github.com/user-attachments/assets/4c111cf2-254d-4830-81c1-30f5023f7932" />

- [x] Pull request is based on the default branch (`3.x` at this time)
- [x] Pull request includes tests for any new/updated operations/facts
- [x] Pull request includes documentation for any new/updated operations/facts
- [x] Tests pass (see `scripts/dev-test.sh`)
- [x] Type checking & code style passes (see `scripts/dev-lint.sh`)

rod7760 committed:
Support metadata for pyinfra plugins.

* Add pyinfra-metadata-schema-1.0.0.json.
* Add pyinfra-metadata.toml with current plugins.
* Add Jinja2 templating to sphinx build.
* Extended build-public-docs.sh to build local docs.
* Added generated index for facts and operations.

DonDebonair commented on docs/metadata.py:
```diff
@@ -0,0 +1,80 @@
+    "security",
+    "storage",
+    "system",
+    "system-ops",
```
Why do we need `system-ops` and `system-facts` next to `system`? For the other tags, we don't have separate `-facts` and `-ops` variants. Whether we're dealing with facts or operations should be clear from the plugin type.

DonDebonair commented on docs/metadata.py:
```diff
@@ -0,0 +1,80 @@
+        raise ValueError("Missing [pyinfra.plugins] section in pyinfra-metadata.toml")
+
+    plugins = []
+    for p in pyinfra_metadata["plugins"]:
```
I wonder if this can be even more concise if we'd use Pydantic for defining the `Tag` and `Plugin` models

DonDebonair commented:
Love this! I would move the `template` folder inside the `docs` folder if possible, because it's only relevant for docs right? And for `metadata.py` I wonder if the other way around makes sense: move it outside of the `docs` folder, because in the future, metadata could be relevant for other things than generating docs.

Maybe we want to create a plugin registry with API (think Ansible Galaxy)

rod7760 commented:
Those both sound reasonable. That template folder is actually from an old implementation. I forgot to delete it. Whoops

Fizzadar commented on pyinfra-metadata.toml:
```diff
@@ -0,0 +1,550 @@
+"$schema" = "./pyinfra-metadata-schema-1.0.0.json"
```
Might be nice to have a script to regenerate this file in-repo since there's so many entries (probably only relevant to pyinfra core itself).

Fizzadar replied:
Eh, might need to put the tags into machine readable format in the docstrings to do that though 🤔

Fizzadar commented on templates/operations.rst.j2:
```diff
@@ -0,0 +1,54 @@
+Operations Index
```
~~Is this file or the existing/modified `operations.rst` being used to generate the docs? Both appear to have identical content...~~ nvm, dupe, to be deleted :)

Fizzadar reviewed (commented):
This is awesome stuff @rod7760! Couple of bits to cleanup Daan already pointed out and good to go I think.

Community context: this is the precursor work that will enable us to generate other pyinfra projects docs inline on docs.pyinfra.com avoiding a disjointed mess of documentation websites.

rod7760 commented:
Resolved the open comments. I moved `docs.metadata` to `pyinfra.api.metadata`. 

Additionally, I now only render the `operations.rst` and `facts.rst` as `jinja2` templates. 
( `src/pyinfra/operations/files.py` contains `jinja2` templates as examples. They caused undefined variable errors.)

We could also explicitly exclude files, but the jinja2 templating felt niche enough to include files explicitly.

Fizzadar committed:
PR feedback
Remove old template dir

Drop system-ops and system-facts

Add pydantic to docs/metadata.py

Move docs/metadata to pyinfra.api.metadata need to
fix docs generation error for missing jinja2 vars

Fix linting and only render subset of docs as jinja2

Update python3 -m http.server command in build-public-docs.sh

# Conflicts:
#	uv.lock

Fizzadar committed:
Restore sidebar links to each op/fact module

Hidden toctree so not shown in the main content/index.

Fizzadar committed:
Fix links to fact/operation modules

Fizzadar committed:
Add under header links between each facts/operations/connectors indices

Fizzadar committed:
Sort the operations/facts index list

Fizzadar committed:
Upgrade docs theme
